File: C/finder_tests.c
#include <setjmp.h>
#include <stdarg.h>
#include <stddef.h>

#include <cmocka.h>
#include <stdlib.h>
#include "finder.h"

static void test_EmptyResultsWhenGivenEmptyList(void** state)
{
    List p;
    p.p = NULL;
    p.size = 0;

    Finder* finder = createFinder(&p);
    F* result = find(finder, One);

    assert_null(result->P1);
    assert_null(result->P2);

    destroyFinder(finder);
}

static void test_EmptyResultsWhenGivenOneFT(void** state)
{
    List p;
    p.p = malloc(sizeof(Thing*));
    p.p[0] = createThing("Sue", 50);
    p.size = 1;

    Finder* finder = createFinder(&p);
    F* result = find(finder, One);

    assert_null(result->P1);
    assert_null(result->P2);

    destroyFinder(finder);
    destroyThing(p.p[0]);
    free(p.p);
}

static void test_OneTwoForTwoFTs(void** state)
{
    List p;
    p.p = malloc(2 * sizeof(Thing*));
    p.p[0] = createThing("Sue", 50);
    p.p[1] = createThing("Greg", 52);
    p.size = 2;

    Finder* finder = createFinder(&p);
    F* result = find(finder, One);

    assert_string_equal(result->P1->name, "Sue");
    assert_string_equal(result->P2->name, "Greg");

    destroyFinder(finder);
    destroyThing(p.p[0]);
    destroyThing(p.p[1]);
    free(p.p);
}

static void test_TwoTwoForTwoFTs(void** state)
{
    List p;
    p.p = malloc(2 * sizeof(Thing*));
    p.p[0] = createThing("Mike", 79);
    p.p[1] = createThing("Greg", 52);
    p.size = 2;

    Finder* finder = createFinder(&p);
    F* result = find(finder, Two);

    assert_string_equal(result->P1->name, "Greg");
    assert_string_equal(result->P2->name, "Mike");

    destroyFinder(finder);
    destroyThing(p.p[0]);
    destroyThing(p.p[1]);
    free(p.p);
}

static void test_TwoTwoForFourFTs(void** state)
{
    List p;
    p.p = malloc(4 * sizeof(Thing*));
    p.p[0] = createThing("Sue", 50);
    p.p[1] = createThing("Sarah", 82);
    p.p[2] = createThing("Mike", 79);
    p.p[3] = createThing("Greg", 52);
    p.size = 4;

    Finder* finder = createFinder(&p);
    F* result = find(finder, Two);

    assert_string_equal(result->P1->name, "Sue");
    assert_string_equal(result->P2->name, "Sarah");

    destroyFinder(finder);
    destroyThing(p.p[0]);
    destroyThing(p.p[1]);
    destroyThing(p.p[2]);
    destroyThing(p.p[3]);
    free(p.p);
}

static void test_OneTwoForFourFTs(void** state)
{
    List p;
    p.p = malloc(4 * sizeof(Thing*));
    p.p[0] = createThing("Sue", 50);
    p.p[1] = createThing("Sarah", 82);
    p.p[2] = createThing("Mike", 79);
    p.p[3] = createThing("Greg", 52);
    p.size = 4;

    Finder* finder = createFinder(&p);
    F* result = find(finder, One);

    assert_string_equal(result->P1->name, "Sue");
    assert_string_equal(result->P2->name, "Greg");

    destroyFinder(finder);
    destroyThing(p.p[0]);
    destroyThing(p.p[1]);
    destroyThing(p.p[2]);
    destroyThing(p.p[3]);
    free(p.p);
}

static const struct CMUnitTest tests[] = {
    cmocka_unit_test(test_EmptyResultsWhenGivenEmptyList),
    cmocka_unit_test(test_EmptyResultsWhenGivenOneFT),
    cmocka_unit_test(test_OneTwoForTwoFTs),
    cmocka_unit_test(test_TwoTwoForTwoFTs),
    cmocka_unit_test(test_TwoTwoForFourFTs),
    cmocka_unit_test(test_OneTwoForFourFTs),
};

int main(void)
{
    return cmocka_run_group_tests(tests, NULL, NULL);
}

File: Python/src/finder_tests.py
#!/usr/bin/env python

from datetime import date
import unittest

from algorithm.ft import FT
from algorithm.finder import Finder
from algorithm.thing import Thing


class FinderTests(unittest.TestCase):

    def setUp(self):
        self.sue = Thing()
        self.greg = Thing()
        self.sarah = Thing()
        self.mike = Thing()
        self.sue.name = "Sue"
        self.sue.birthdate = date(1950, 1, 1)
        self.greg.name = "Greg"
        self.greg.birthdate = date(1952, 6, 1)
        self.sarah.name = "Sarah"
        self.sarah.birthdate = date(1982, 1, 1)
        self.mike.name = "Mike"
        self.mike.birthdate = date(1979, 1, 1)

    def test_returns_empty_results_when_given_empty_list(self):
        array = []
        finder = Finder(array)

        result = finder.find(FT.One)

        assert result.p1 is None
        assert result.p2 is None

    def test_returns_empty_results_when_given_one_ft(self):
        array = []
        array.append(self.sue)

        finder = Finder(array)

        result = finder.find(FT.One)

        assert result.p1 is None
        assert result.p2 is None

    def test_returns_one_two_for_two_fts(self):
        array = []
        array.append(self.sue)
        array.append(self.greg)
        finder = Finder(array)

        result = finder.find(FT.One)

        assert self.sue == result.p1
        assert self.greg == result.p2

    def test_returns_two_two_for_two_fts(self):
        array = []
        array.append(self.mike)
        array.append(self.greg)
        finder = Finder(array)

        result = finder.find(FT.Two)

        assert self.greg == result.p1
        assert self.mike == result.p2

    def test_returns_two_two_for_four_fts(self):
        array = []
        array.append(self.sue)
        array.append(self.sarah)
        array.append(self.mike)
        array.append(self.greg)
        finder = Finder(array)

        result = finder.find(FT.Two)

        assert self.sue == result.p1
        assert self.sarah == result.p2

    def test_returns_one_two_for_four_fts(self):
        array = []
        array.append(self.sue)
        array.append(self.sarah)
        array.append(self.mike)
        array.append(self.greg)
        finder = Finder(array)

        result = finder.find(FT.One)

        assert self.sue == result.p1
        assert self.greg == result.p2


if __name__ == "__main__":
    unittest.main()
